packages/react-resizable-panels/src/PanelGroup.test.tsx

```diff
@@ -130,6 +130,110 @@ describe("PanelGroup", () => {
     expect(rightPanelElement?.getAttribute("data-panel-size")).toBe("40.0");
   });
 
+  it("should rerender properly after the entire layout is changed", () => {
+    const layouts = [
+      {
+        id: "parent",
+        direction: "vertical",
+        children: [
+          {
+            id: "foo",
+            props: {
+              area: {
+                min: 30,
+              },
+            },
+          },
+          {
+            id: "bar",
+            props: {
+              area: {
+                min: 70,
+              },
+            },
+          },
+        ],
+      },
+      {
+        id: "parent",
+        direction: "horizontal",
+        children: [
+          {
+            id: "baz",
+          },
+          {
+            id: "bon",
+            props: {
+              area: {
+                max: 25,
+              },
+            },
+          },
+        ],
+      },
+    ];
+
+    // Helper components
+    const isContainer = (node: any) => "children" in node;
+
+    const CustomPanel = ({ panel }: any) => {
+      return <div>{panel.id}</div>;
+    };
+
+    const Container = ({ container: { id, direction, children } }: any) => {
+      return (
+        <PanelGroup id={id} direction={direction} key={id}>
+          <Panel
+            id={children[0].id}
+            defaultSize={children[0].props?.area?.min}
+            minSize={children[0].props?.area?.min}
+            maxSize={children[0].props?.area?.max}
+          >
+            {isContainer(children[0]) ? (
+              <Container container={children[0]} />
+            ) : (
+              <CustomPanel panel={children[0]} />
+            )}
+          </Panel>
+          <PanelResizeHandle
+            style={{ height: "2px", backgroundColor: "red" }}
+          />
+          {children.slice(1).map((child: any, index: any) => {
+            return (
+              <Fragment key={child.id}>
+                <Panel
+                  id={child.id}
+                  defaultSize={child.props?.area?.min}
+                  minSize={child.props?.area?.min}
+                  maxSize={child.props?.area?.max}
+                >
+                  {isContainer(child) ? (
+                    <Container container={child} />
+                  ) : (
+                    <CustomPanel panel={child} />
+                  )}
+                </Panel>
+                {index < children.length - 2 && (
+                  <PanelResizeHandle
+                    style={{ height: "2px", backgroundColor: "red" }}
+                  />
+                )}
+              </Fragment>
+            );
+          })}
+        </PanelGroup>
+      );
+    };
+
+    act(() => {
+      root.render(<Container container={layouts[0]} />);
+    });
+
+    act(() => {
+      root.render(<Container container={layouts[1]} />);
+    });
+  });
```

Comment:
I _think_ this test could be reduced to the following (so that it's easier to read)?
```tsx
  it("should rerender properly after the entire layout is changed", () => {
    act(() => {
      root.render(
        <PanelGroup id="parent" direction="vertical">
          <Panel id="foo" minSize={30} order={0} />
          <PanelResizeHandle />
          <Panel id="bar" minSize={70} order={1} />
        </PanelGroup>
      );
    });

    act(() => {
      root.render(
        <PanelGroup id="parent" direction="horizontal">
          <Panel id="baz" order={0} />
          <PanelResizeHandle />
          <Panel id="bar" maxSize={25} order={1} />
        </PanelGroup>
      );
    });
  });
```
Note that I added the missing `order` prop (that was being warned about)

Reply:
I agree, I didn't knew what was the root cause so I added my entire sandbox project. I'll revise it

Reply:
No need. I'm looking into this already. If I decide I need to add new tests to cover this area (likely) I don't mind writing them.

Reply:
I've been able to trim it down a little farther to this failing case:
```tsx
it("should rerender properly after layout is changed", () => {
    act(() => {
      root.render(
        <PanelGroup direction="vertical">
          <Panel id="foo" minSize={30} order={0} />
          <PanelResizeHandle />
          <Panel id="bar" minSize={70} order={1} />
        </PanelGroup>
      );
    });

    act(() => {
      root.render(
        <PanelGroup direction="vertical">
          <Panel id="bar" minSize={70} order={0} />
        </PanelGroup>
      );
    });
```

Reply:
I _think_ this change (and test) covers it
```diff
diff --git a/packages/react-resizable-panels/src/PanelGroup.test.tsx b/packages/react-resizable-panels/src/PanelGroup.test.tsx
index c6f747a..abb12ce 100644
--- a/packages/react-resizable-panels/src/PanelGroup.test.tsx
+++ b/packages/react-resizable-panels/src/PanelGroup.test.tsx
@@ -130,6 +130,35 @@ describe("PanelGroup", () => {
     expect(rightPanelElement?.getAttribute("data-panel-size")).toBe("40.0");
   });
 
+  // github.com/bvaughn/react-resizable-panels/issues/303
+  it("should recalculate layout after panels are changed", () => {
+    let mostRecentLayout: number[] | null = null;
+
+    const onLayout = (layout: number[]) => {
+      mostRecentLayout = layout;
+    };
+
+    act(() => {
+      root.render(
+        <PanelGroup direction="vertical" onLayout={onLayout}>
+          <Panel id="foo" minSize={30} order={0} />
+          <PanelResizeHandle />
+          <Panel id="bar" minSize={70} order={1} />
+        </PanelGroup>
+      );
+    });
+    expect(mostRecentLayout).toEqual([30, 70]);
+
+    act(() => {
+      root.render(
+        <PanelGroup direction="vertical" onLayout={onLayout}>
+          <Panel id="bar" minSize={70} order={0} />
+        </PanelGroup>
+      );
+    });
+    expect(mostRecentLayout).toEqual([100]);
+  });
+
   describe("imperative handle API", () => {
     it("should report the most recently rendered group id", () => {
       const ref = createRef<ImperativePanelGroupHandle>();
diff --git a/packages/react-resizable-panels/src/PanelGroup.ts b/packages/react-resizable-panels/src/PanelGroup.ts
index 469d0e6..ce167c7 100644
--- a/packages/react-resizable-panels/src/PanelGroup.ts
+++ b/packages/react-resizable-panels/src/PanelGroup.ts
@@ -768,10 +768,10 @@ function PanelGroupWithForwardedRef({
         panelData,
         layout
       );
-      assert(
-        prevPanelSize != null,
-        `Previous panel size not found for panel "${panelData.id}"`
-      );
+      if (prevPanelSize == null) {
+        // It's possible that the panels in this group have changed since the last render
+        return;
+      }
 
       if (
         prevCollapsible &&
```